redhat-developer/rhdh-plugins
# feat(lightspeed): add permission support in lightspeed plugins

Tickets:

https://issues.redhat.com/browse/RHDHPAI-33
https://issues.redhat.com/browse/RHDHPAI-34
https://issues.redhat.com/browse/RHDHPAI-35

## Lightspeed Permission support


Add permission support for lightspeed plugins.

This PR adds three basic permissions `lightspeed.conversations.read`, `lightspeed.conversations.create` and  `lightspeed.conversations.delete`. These permission are used to secure the UI and API endpoints.

**Note:**

To access lightspeed UI, then logged in user should be granted with below permissions.
 
`lightspeed.conversations.read` 
`lightspeed.conversations.create`

API endpoints:

`lightspeed.conversations.read` - all `GET` requests are gated by this permission
`lightspeed.conversations.create` - all `POST` requests are gated by this permission
`lightspeed.conversations.delete` - all `DELETE` requests are gated by this permission


## How to test this changes?



1. **Install and configure [rbac-backend](https://github.com/backstage/community-plugins/blob/main/workspaces/rbac/plugins/rbac-backend/README.md#installation)** (Prerequiste)
2. create a `rbac-policy.csv` file with below permission policies.

       
        p, role:default/team_a, lightspeed.conversations.read, read, allow
        p, role:default/team_a, lightspeed.conversations.create, create, allow
        p, role:default/team_a, lightspeed.conversations.delete, delete, allow
        
        g, user:default/<your-user-name>, role:default/team_a
         

3.  enable rbac and wire the policy file in the `app-config.yam`l file

        permission:
          enabled: true
          rbac:
            policies-csv-file: /your/absolute/path/rbac-policy.csv
            policyFileReload: true

 - run the application and modify the permission policy from `allow` to `deny` to see the permission required alert or Unauthorized error while accessing endpoints.
 
## Screenshots




If one of the permissions (`lightspeed.conversations.read` or `lightspeed.conversations.create`) is denied for a user, then UI will show the missing permissions error state.


![image](https://github.com/user-attachments/assets/cedb9264-4a47-4596-b194-ea2b880bff26)

Delete button is disabled when `lightspeed.conversations.delete` is denied.

![image](https://github.com/user-attachments/assets/f92617d3-1260-4706-a5f0-1913c76ddb96)


**API endpoints**

![conversation_access_denied](https://github.com/user-attachments/assets/5d3ce010-3686-4c6d-a809-0332d61be728)
![models_api_access_denied](https://github.com/user-attachments/assets/1394dc88-40dc-4884-85d3-b5938ac3e8b8)

## Unit test

#### :heavy_check_mark: Checklist
